ascii/valid_asm.go

```diff
@@ -8,73 +8,40 @@ import (
 )
 
 func main() {
-	TEXT("validPrint16", NOSPLIT, "func(p *byte, n uintptr) int")
-	Doc("Validates that the string only contains printable ASCII characters.")
+	TEXT("valid16", NOSPLIT, "func(p *byte, n uintptr) int")
+	Doc("Validates that the string only contains ASCII characters.")
 
 	p := Load(Param("p"), GP64())
 	n := Load(Param("n"), GP64())
 	r := GP64()
 	MOVQ(U64(0), r)
 
-	Comment("Initialize 128 bits registers.")
-	min := GP64()
-	max := GP64()
-	MOVQ(U64(0x1919191919191919), min)
-	MOVQ(U64(0x7E7E7E7E7E7E7E7E), max)
-
-	minXMM := XMM()
-	PINSRQ(Imm(0), min, minXMM)
-	PINSRQ(Imm(1), min, minXMM)
-
-	maxXMM := XMM()
-	PINSRQ(Imm(0), max, maxXMM)
-	PINSRQ(Imm(1), max, maxXMM)
-
-	xmm0 := XMM()
-	xmm1 := XMM()
 	msk0 := GP32()
+	msk1 := GP32()
 
-	CMPQ(n, Imm(2)) // skip YMM register initialization if there are less than 32 bytes
-	JL(LabelRef("loop16"))
-
-	Comment("Initialize 256 bits registers.")
-	minYMM := YMM()
-	VPBROADCASTQ(minXMM, minYMM)
-
-	maxYMM := YMM()
-	VPBROADCASTQ(maxXMM, maxYMM)
-
+	xmm0 := XMM()
 	ymm0 := YMM()
 	ymm1 := YMM()
-	ymm2 := YMM()
-	ymm3 := YMM()
-	ymm4 := YMM()
-	ymm5 := YMM()
-	ymm6 := YMM()
-	ymm7 := YMM()
-	ymm8 := YMM()
 
 	Label("loop64")
-	Comment("Unroll two iterations of the loop operating on 32 bytes chunks.")
 	CMPQ(n, Imm(4))
 	JL(LabelRef("loop32"))
 
+	// TODO: it would be nice to be able to use VPMOVB2M here, but it appears
+	// that the current version of avo hasn't yet released support for the
+	// AVX.512 instruction set.
 	VMOVUPS(Mem{Base: p}, ymm0)
 	VMOVUPS((Mem{Base: p}).Offset(32), ymm1)
-	VPCMPGTB(minYMM, ymm0, ymm2) // A = bytes that are greater than the min-1 (i.e. valid at lower end)
-	VPCMPGTB(maxYMM, ymm0, ymm3) // B = bytes that are greater than the max (i.e. invalid at upper end)
-	VPANDN(ymm2, ymm3, ymm4)     // A & ~B mask should be full unless there's an invalid byte
-	VPCMPGTB(minYMM, ymm1, ymm5) // compute the same for the next 32 bytes
-	VPCMPGTB(maxYMM, ymm1, ymm6)
-	VPANDN(ymm5, ymm6, ymm7)
-	VPAND(ymm4, ymm7, ymm8)      // combine masks
-	VPMOVMSKB(ymm8, msk0)
-	XORL(U32(0xFFFFFFFF), msk0)  // check for a zero somewhere
+
+	VPMOVMSKB(ymm0, msk0)
+	VPMOVMSKB(ymm1, msk1)
+	ORL(msk1, msk0)
+	CMPL(msk0, Imm(0))
```

Comment:
This is a bit faster for me:

```golang
VMOVUPS(Mem{Base: p}, ymm0)
VPOR((Mem{Base: p}).Offset(32), ymm0, ymm1)
VPMOVMSKB(ymm1, msk0)
CMPL(msk0, Imm(0))
```

```
Valid/12345678901234567...-4          70.3GB/s ± 0%  88.4GB/s ± 0%  +25.64%  (p=0.008 n=5+5)
```

Reply:
ah nice! I'll add that in

Reply:
```
$ benchstat /tmp/bench{5,6}
name                           old time/op    new time/op     delta
Valid/a_string_of_16B.           7.76ns ± 2%     7.44ns ± 4%    -4.15%  (p=0.040 n=5+5)
Valid/an_invalid_string...       8.47ns ± 2%     6.94ns ± 6%   -18.11%  (p=0.008 n=5+5)
Valid/some_kind_of_long...       10.3ns ± 2%      8.4ns ± 2%   -17.86%  (p=0.008 n=5+5)
Valid/some_kind_of_long...#01    8.87ns ± 2%     5.78ns ± 3%   -34.83%  (p=0.008 n=5+5)
Valid/12345678901234567...        559ns ± 6%      103ns ± 3%   -81.50%  (p=0.008 n=5+5)

name                           old speed      new speed       delta
Valid/a_string_of_16B.         2.06GB/s ± 2%   2.15GB/s ± 4%    +4.38%  (p=0.032 n=5+5)
Valid/an_invalid_string...     3.90GB/s ± 2%   4.76GB/s ± 6%   +22.18%  (p=0.008 n=5+5)
Valid/some_kind_of_long...     5.06GB/s ± 2%   6.17GB/s ± 2%   +22.09%  (p=0.008 n=5+5)
Valid/some_kind_of_long...#01  7.22GB/s ± 2%  11.08GB/s ± 3%   +53.48%  (p=0.008 n=5+5)
Valid/12345678901234567...     17.9GB/s ± 6%   96.8GB/s ± 3%  +440.52%  (p=0.008 n=5+5)
```

🚀

Reply:
Added in https://github.com/segmentio/encoding/commit/2c5ebb4ec4eebcd65fce51dbfae445df53a6a6d7

Reply:
https://github.com/segmentio/encoding/pull/70